### ISSUE TITLE
vagrant/client: the "installing software" step is missing

### ISSUE THREAD

**Issue by rjeczalik** [koding/koding#9040]
When building vagrant stack the "Installing software" step after "Checking VM connections" is missing, and it used to be there.

## Expected Behavior

Modal should tail on /var/log/cloud-init-output.log

## Current Behavior

It does tail.

## Possible Solution

It should tail.

## Steps to Reproduce (for bugs)

1. Build a vagrant stack.

## Your Environment

team.koding.com

**Comment by gokmen** [comment]
@alex-ionochkin can u check this issue and make it work with vagrant provider as well like we did for aws provider? Thanks!

**Comment by alex-ionochkin** [comment]
Actually `Installing software` step appears after `Checking VM Connections` but it's visible for a short period of time and then it quickly changes to success screen. The reason of it is that once VM is ready and `/var/log/cloud-init-output.log` is loaded logs file already contains `_KD_DONE_` which is a signal to show success screen on the modal. I tried different `sleep` values in `user_data` section of stack template to make a software installation longer but every time I got the same result - `Installing software` step quickly disappeared.
It looks like `user_data` instructions are done before modal `apply finished` event comes from the kloud. For example, if I run a simple vagrant stack without any delay:
```
resource:
  vagrant_instance:
    localvm:
      cpus: 2
      memory: 2048
      box: ubuntu/trusty64
      user_data: |-
        sudo apt-get install sl -y
        touch /tmp/${var.koding_user_username}.txt
```
it takes ~2 min. If I run the same template with 2 mins of `sleep`:
```
resource:
  vagrant_instance:
    localvm:
      cpus: 2
      memory: 2048
      box: ubuntu/trusty64
      user_data: |-
        sudo apt-get install sl -y
        touch /tmp/${var.koding_user_username}.txt
        sleep 120
```
it takes ~4min and most of time is taken by `Building stack resources` step.

@rjeczalik, could you please check on the kloud if machine already has `_KD_DONE_` in `/var/log/cloud-init-output.log` before kloud sends `apply finished` stack event to the client? If it's so, modal can't show `Installing Software` step since it's already done

**Comment by rjeczalik** [comment]
> could you please check on the kloud if machine already has _KD_DONE_ in /var/log/cloud-init-output.log before kloud sends apply finished stack event to the client?

_KD_DONE_ can be in the /var/log/cloud-init-output.log even when `Checking connectivity` step is not done yet.

The way it works is that cloud-init installs klient and runs user-data.sh asynchronously to the events, so you can't make any assumptions about the events order here. If installing software is already done it should display that installing software is already done, I guess.

Installing klient and software is one timeline of events and kloud observing that is another one. Terraforms building infrastructure is a synchronous event and kloud waits until Terraform is done (this is the "Building..." step). When Terraformer is done, then the actual cloud provider is executing cloud-init - neither kloud nor client knows when it's done.

What kloud does instead is to poll kontrol waiting for a klient to register (this is the `Checking connectivity` step). When that step is done `apply finished` event is emitted. There's is no guarantee, as you already discovered, kloud will observe registered klient before the user-script.sh is done executing.

**Comment by rjeczalik** [comment]
@gokmen & @sibblegp do we want to display logs if user-data execution is fast and there's nothing to tail on?

**Comment by alex-ionochkin** [comment]
As we discussed with @rjeczalik, the reason of the problem is that for vagrant stacks `apply finished` event comes always when user-data execution is done and no matter whether it's fast or long. In current implementation of `ResourceStateModal` it shows `Installing software` process after it gets `apply finished` event and before `_KD_DONE_` appears in logs. Since `_KD_DONE_` is already in logs when `apply finished` event comes, `Installing software` and logs disappears very quickly since success screen replaces them

**Comment by gokmen** [comment]
So, what's the plan to be able to make it like in aws stacks? are we going to change the FE or do @rjeczalik need to change vagrant provider?

**Comment by rjeczalik** [comment]
@gokmen I can change, but I'm not sure the assumption we make here will stand for other providers - that after `apply finished` event user-data is still being executed.

**Comment by sibblegp** [comment]
Clearly, whatever event that signals that "Connecting to VMs....." is done is not being received so that "Apply Finished" can be displayed and then "Installing Software" can be displayed prior to all of the software being installed.  If the software installation is quick, there is no need to tail cloud-init-output.log.  But most software installs are long and we want to tail them and right now that isn't happening.  We just need to figure out what signal from vagrant/kloud (I think?) means that we can move on from "Connecting to VMs...." and then start tailing cloud-init-output.log until we get a _KD_DONE_. on Vagrant.

**Comment by gokmen** [comment]
@rjeczalik I think klient installation should be done before everything and then we should start executing the `user_data` for all providers and I think this will solve most of the issues there even though I'm not sure how easy to implement it in that way.
